File: bindings/c/src/paint.cpp
// Copyright (c) 2025 Daniel T. McGinnis
// SPDX-License-Identifier: MIT

#include <nux/paint.h>
#include "handles.hpp"

NUXC_EXTERNC
nux_gradient_stop_t*
nux_gradient_stop_create (
    float position,
    nux_gvector4d_t* color
) {
    return new nux_gradient_stop_t{
        nux::gradient_stop{position, *nux_impl_map(color)}
    };
}

NUXC_EXTERNC
void
nux_gradient_stop_destroy (
    nux_gradient_stop_t* gradient_stop
) {
    delete gradient_stop;
}

NUXC_EXTERNC
void
nux_paint_destroy (
    nux_paint_t* paint
) {
    delete paint;
}

NUXC_EXTERNC
nux_paint_t*
nux_paint_build_solid_color (
    nux_gvector4d_t* color,
    float global_opacity
) {
    return new nux_paint_t{
        nux::paint::build_solid_color(*nux_impl_map(color), global_opacity)
    };
}

NUXC_EXTERNC
nux_paint_t*
nux_paint_build_linear_gradient (
    nux_gvector2d_t* start,
    nux_gvector2d_t* end,
    nux_gradient_stop_t** stops,
    int stop_count,
    float global_opacity
) {
    return new nux_paint_t{
        nux::paint::build_linear_gradient(
            *nux_impl_map(start),
            *nux_impl_map(end),
            nux_impl_map<nux::gradient_stop>(stops, stop_count),
            global_opacity
        )
    };
}

NUXC_EXTERNC
nux_paint_t*
nux_paint_build_radial_gradient (
    nux_gvector2d_t* center,
    float radius,
    nux_gradient_stop_t** stops,
    int stop_count,
    float global_opacity
) {
    return new nux_paint_t{
        nux::paint::build_radial_gradient(
            *nux_impl_map(center),
            radius,
            nux_impl_map<nux::gradient_stop>(stops, stop_count),
            global_opacity
        )
    };
}

File: bindings/rs/nux/src/paint.rs
// Copyright (c) 2025 Daniel T. McGinnis
// SPDX-License-Identifier: MIT

use std::ffi::*;
use crate::bindings::*;
use crate::gvector2d::*;
use crate::gvector4d::*;
use crate::gradient_stop::*;

#[derive(Clone)]
pub struct Paint {
    pub(crate) native_handle: *mut nux_paint_t
}

impl Paint {
    pub fn build_solid_color (
        color: GVector4D,
        global_opacity: f32
    ) -> Self {
        return Self {
            native_handle: unsafe {
                nux_paint_build_solid_color(
                    color.native_handle,
                    global_opacity
                )
            }
        };
    }

    pub fn build_linear_gradient (
        start: GVector2D,
        end: GVector2D,
        stops: &Vec<GradientStop>,
        global_opacity: f32
    ) -> Self {
        return Self {
            native_handle: unsafe {
                nux_paint_build_linear_gradient(
                    start.native_handle,
                    end.native_handle,
                    stops.as_ptr() as *mut *mut nux_gradient_stop_t,
                    stops.len() as c_int,
                    global_opacity
                )
            }
        };
    }

    pub fn build_radial_gradient (
        center: GVector2D,
        radius: f32,
        stops: &Vec<GradientStop>,
        global_opacity: f32
    ) -> Self {
        return Self {
            native_handle: unsafe {
                nux_paint_build_radial_gradient(
                    center.native_handle,
                    radius,
                    stops.as_ptr() as *mut *mut nux_gradient_stop_t,
                    stops.len() as c_int,
                    global_opacity
                )
            }
        };
    }
}

impl Drop for Paint {
    fn drop (
        &mut self
    ) {
        unsafe {
            nux_paint_destroy(self.native_handle)
        };
    }
}
